### PR TITLE
ARROW-2401 Support filters on Hive partitioned Parquet files

### PR DESCRIPTION
**Pull request by jneuff**
This PR enables filtering based on a Hive partitioned directory structure of Parquet files. Only `=` and `!=` are supported as filters.

### CODE REVIEW

**Review by xhochy — COMMENTED** [review]
+1, LGTM. 

Retrigged the Travis build that timeouted, will merge on green build.

File: python/pyarrow/parquet.py [hunk 1]
```diff
@@ -849,6 +857,31 @@ def open_file(path, meta=None):
...
+    def _filter(self, filters):
+        def filter_accepts_partition(part_key, filter, level):
+            p_column, p_value_index = part_key
+            f_column, op, f_value = filter
+            if p_column != f_column:
+                return True
+
+            f_value_index = self.partitions.get_index(level, p_column,
+                                                      str(f_value))
+            if op == "=":
```
Comment (xhochy) [inline python/pyarrow/parquet.py]: Clarified offline with @jneuff that we will refactor `self.partitions.get_index` a bit to support more operation than the two above.

**Comment by gsakkis** [comment]
Currently the filters allow effectively at most one filter per partition level. It would be useful to extend them to support a logical `OR` of partition values, especially given the lack of inequality operators. Suggested syntax:
```
filters=[
    ('foo', '=', [1, 5, 7]), 
    ('bar', '!=', ['x', 'y', 'z'])
]
```
Basically check if the value (third item) is a list/tuple (or perhaps any iterable except for strings) and treat it as an OR.

Note: the `!=` is not strictly necessary (it can be expressed equivalently as multiple AND-ed filters) but it's good to have for convenience and consistency. 

I could work on a patch if you agree.

**Comment by wesm** [comment]
+1. It would be great to define a more extensible API in general (for which the syntax defined here is syntactic sugar). You could imagine the current partition keys being passed as a list of `(partition name, partition value)` tuples, and a function could decide whether or not to accept the partition leaf. 

Is there a JIRA yet about any partition filtering generalizations?

**Comment by xhochy** [comment]
I would prefer when we extend the current API to allow lists of lists of predicates. These predicates should be in disjunctive normal form (DNF).

This means that the innermost tuple describe a single column predicate. These inner predicate make are all combined with a conjunction (AND) into a larger predicate. The most outer list then combines all predicates with a disjunction (OR). By this, we should be able to express all kinds of predicates that are possible using boolean logic.

I have running code for this which I could submit once I find the time to adapt it to the ParquetDataset API.

**Pull request closed without merging**